### PR TITLE
fix: correct label masking in legacychat_tempalte

### PR DESCRIPTION
**Pull request by ved1beta**
# Description
the legacy chat template path was incorrectly masking all tokens
Replace global prompt length masking with per-turn role-based masking

## Motivation and Context
#3088

# test 
config in issue works with qwen3 0.6B and given data set 
``` {"messages": [
  {"role": "user", "content": "Hello"},
  {"role": "assistant", "content": "Hi there!"},
  {"role": "user", "content": "How are you?"},
  {"role": "assistant", "content": "<think>I should respond positively</think>I'm doing great, thanks for asking!"}
]}
{"messages": [
  {"role": "user", "content": "What's 2+2?"},
  {"role": "assistant", "content": "2 + 2 equals 4."},
  {"role": "user", "content": "What about 3+3?"},
  {"role": "assistant", "content": "<think>This is simple addition</think>3 + 3 equals 6."}
]}
{"messages": [
  {"role": "user", "content": "Tell me about cats"},
  {"role": "assistant", "content": "Cats are wonderful pets!"},
  {"role": "user", "content": "Do they like water?"},
  {"role": "assistant", "content": "<think>Most cats don't like water, but there are exceptions</think>Most cats don't enjoy water, but some breeds like Maine Coons actually like swimming!"}
]}
{"messages": [
  {"role": "user", "content": "What's the capital of France?"},
  {"role": "assistant", "content": "<think>This is a straightforward geography question</think>The capital of France is Paris."}
]}
{"messages": [
  {"role": "user", "content": "Good morning!"},
  {"role": "assistant", "content": "Good morning! How can I help you today?"},
  {"role": "user", "content": "I need help with my homework"},
  {"role": "assistant", "content": "I'd be happy to help!"},
  {"role": "user", "content": "It's about physics"},
  {"role": "assistant", "content": "<think>Physics can be tricky, I should ask what specific topic they need help with</think>Physics is a fascinating subject! What specific topic are you working on?"}
]} ```



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed label masking when training without user inputs so all user turns in multi-turn conversations are masked, preventing user text from influencing the loss and improving training stability and fine-tuning alignment. No public API changes.

* **Tests**
  * Updated prompt tokenization/masking tests to reflect the new per-turn masking behavior and token-level alignment in multi-turn prompts.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

### CODE REVIEW

**Comment by winglian** [comment]
I thought this was already well tested and behaved correctly. What's the failure mode that you're fixing? (we should probably add a unit test in that case if it's something we don't have a test for)

**Comment by ved1beta** [comment]
chat  template was incorrectly masking tokes mentioned in #3088 , will need to change existing test i guess the one which is failing
